### ISSUE TITLE
`fsync` (`fdatasync`) WAL xlog asynchronously to amortize syncing overhead

### ISSUE THREAD

**Issue by CuriousGeorgiy** [tarantool/tarantool#10785]
Currently, we open the WAL files with `O_SYNC` for data integrity:
https://github.com/tarantool/tarantool/blob/8806b7b561d58d1cef79e6c9c723827f4ce67ed8/src/box/wal.c#L441-L442

This means that each WAL write is coupled with a filesystem synchronization.

Instead, we could asynchronously `fsync` (`fdatasync`) the WAL files in a separate WAL thread fiber, amortizing the syncing overhead. To ensure durability guarantees, the transactions depending on the sync would be acknowledged by this fiber.

This would prevent the WAL thread from stalling on filesystem synchronization and potentially improve its throughput.

**Comment by Totktonada** [comment]
Isn't it what tarantool performs by default (`wal_mode = 'write'`)?

The quoted code seems related to `wal_mode = 'fsync'`.

Once we observed a performance degradation that is related to a mistake, which effectively makes tarantool writing xlog files without `fdatasync` (just like it is not supported). It was fixed in cd9cc4c5057581970db4f8107c2a86ff0b24f4e2 and you can see performance measurements in the linked issue.

**Comment by CuriousGeorgiy** [comment]
@Totktonada 

> Isn't it what tarantool performs by default (`wal_mode = 'write'`)?

No, the `sync_is_async` option you mentioned is used only in `xlog_sync` for calling `fsync` on the xlog file before closing it — which is, AFAIC, actually redundant for the WAL xlog files, since they are opend with `O_SYNC`, which already gives the required file integrity guarantees [^1]:
> Write operations on the file will complete according to the requirements of synchronized I/O file integrity completion (by contrast with the synchronized I/O data integrity completion provided by O_DSYNC.)
>
> By the time [write(2)](https://man7.org/linux/man-pages/man2/write.2.html) (or similar) returns, the output data and associated file metadata have been transferred to the underlying hardware (i.e., as though each [write(2)](https://man7.org/linux/man-pages/man2/write.2.html) was followed by a call to [fsync(2)](https://man7.org/linux/man-pages/man2/fsync.2.html)).

[^1]: https://man7.org/linux/man-pages/man2/open.2.html
